### ISSUE TITLE
Fl_Terminal control character handling

### ISSUE THREAD

**Issue by Albrecht-S** [fltk/fltk#950]
I believe the handling of the `Escape` character alone in Fl_Terminal's output is inconsistent and could be improved. Tests were done in current Git master (05a95e146b2429154b360030d098f0dae5555706).

The following demo program shows the inconsistency:
```C++
#include <FL/Fl_Double_Window.H>
#include <FL/Fl_Terminal.H>

int main(int argc, char **argv) {
  auto win = new Fl_Double_Window(500, 400, "Fl_Terminal control character test");
  auto tty = new Fl_Terminal(0, 0, win->w(), win->h());
  win->end();
  win->resizable(tty);
  win->show(argc, argv);
  tty->show_unknown(true);
  tty->printf("%s:\n", win->label());
  char buf[40];
  int len;
  for (int i = 0; i < 32; i++) {
    len = sprintf(buf, "   ASCII(%2d) = '%c'.\n", i, i);
    tty->append(buf, len);
  }
  tty->append("Intentional typo (']') in 2nd Esc sequence:\n");
  len = sprintf(buf, "AB%c[7mCD%c]mEF%c[mGH\n", 27, 27, 27);
  tty->append(buf, len);
  tty->append("Suggested output would be like:\n");
  len = sprintf(buf, "AB%c[7mCD%c]mEF%c[mGH\n", 27, 6, 27);
  tty->append(buf, len);
  tty->append("Done.\n");
  return Fl::run();
}
```

IMO everything but the `Esc` character is handled as expected, particularly `BS`, `LF`, and `CR` are interpreted as intended, see output of the demo program.

The problem with `Esc` is:
- it doesn't output anything for the invalid Esc sequence
- it consumes (suppresses) the next character, in this case "`'`".

I suggest to output the "unknown" char plus the next character if an invalid Escape sequence is detected and not to suppress the next character (the one that invalidates it).

A possible solution would be to output every character of the invalid Esc sequence (the internal buffer) as a single character, replacing `Esc` with "`¿`" etc.. However, the processing should take into account whether `show_unknown()` is true or false. No matter what its value is, the final `'` (in the demo above) should be visible.

See an example and a proposed "solution" at the end of the demo program.

**Comment by Albrecht-S** [comment]
Screenshot:
![image](https://github.com/fltk/fltk/assets/26235500/c028a851-acf7-4a8f-8349-6126d78cf647)

**Comment by Albrecht-S** [comment]
Screenshot with `show_unknown(false)`:
![image](https://github.com/fltk/fltk/assets/26235500/eddab8a5-5c62-4944-8a35-7cbdd895361a)

**Comment by erco77** [comment]
Actually, I think it's important the terminal let bad ESC sequences through, or you'll never know there's a problem. In the old days you'd read the manual for physical terminals and use only the codes they recommended, and if you send garbage to the terminal (such as cat'ing a binary file), the behavior's "undefined", which ranges from just passing thru the garbage to hanging the tty firmware.

I think letting the characters past the ESC sequence is informative in determining that there is an issue; it might be fatal to the operation of the program if ESC codes are malformed, so it should be clear something is wrong, and not try to second guess what the user should see. esp. considering all the possible ways an ESC code can be mangled; you can't tell what to ignore and what is non-esc sequence data.

**Comment by Albrecht-S** [comment]
I believe I agree with Greg (if I understand what he wrote correctly): display everything we can to let the user get some clues what's going on.

There might be a few exceptions: if there is an Esc sequence we can't support in `Fl_Terminal` for some reasons it **might** be useful to suppress the entire sequence (but these cases should be documented).

Example: I'm thinking of the Esc sequences that switch to double height (and maybe double width) characters, displaying either the top half or the bottom half of the characters. In such a case the program would send two identical lines of text with two different preceding Esc sequences (first line "top half", second line "bottom half"). If we can't display the text accordingly we should document that these Esc sequences are recognized but ignored. Same for similar **valid** Esc sequences that can't be displayed in `Fl_Terminal`.

**Comment by MatthiasWM** [comment]
I don't mind either way.

**Comment by Albrecht-S** [comment]
@erco77 Greg, what's the status of this issue? Did you find time to work on it, or have you been too busy? Can we help, maybe?

**Comment by Albrecht-S** [comment]
> It doesn't seem like something holding up 1.4.0, is it?

No, it wouldn't. We could always do it later, I don't think that ABI changes would be necessary.

**Comment by erco77** [comment]
OK, took a shot at addressing this in f825fca.
![image](https://github.com/user-attachments/assets/0f0bce19-ac12-4255-b952-896de27e9d20)

**Comment by Albrecht-S** [comment]
@erco77 Thanks, this seems to work well, at least for the given demo program (tested here too).

I suggest that the docs make clear that setting `void error_char(const char* val)` does not copy the string (it stores only a pointer). It should be mentioned that the string must not be `free`'d as long as it is used, similar to `Fl_Widget::label()` etc..

And for completeness and consistency: correct spelling is `UTF-8`, not `utf8`. This should be fixed at least in the docs of the public API.

**Comment by Albrecht-S** [comment]
@erco77 Any news on this issue? I'm not sure what your :+1: above means:
- Yes, thanks, I (@erco77) will do it, or
- Thanks, I'd like you (@Albrecht-S) to do it

Please let me know if you intend to work further on it.

**Comment by erco77** [comment]
Ya, still on it.. just been slow to commit because of life stuff distracting me.

**Comment by Albrecht-S** [comment]
@erco77 Ping ;-)

**Comment by erco77** [comment]
Yes, noted -- I retooled one of the internal methods while adding the utf8 doc changes, and just need to double check it. Hard to get back up on the horse when you've been away from it for a while..

**Comment by erco77** [comment]
Latest commit fa1aaa7 should solve this issue; your demo now runs as:

<img width="431" height="127" alt="Image" src="https://github.com/user-attachments/assets/7c3e9cd1-2ea2-48e8-9eae-f8c9122eecd6" />

I also added some diagnostics in the terminal application which tests a few invalid xterm sequences as well as a handful of bad utf-8 sequences (which currently don't detect properly, pending my suggs on fltk.coredev for updating fl_utf8len() to detect these cases).

Closing this, unless you see the need to reopen.
